src/js/plugin.js

```diff
@@ -0,0 +1,436 @@
+ *
+ * @private
+ * @param  {Player} player
+ *         A Video.js player instance.
+ *
+ * @param  {string} name
+ *         The name of a plugin.
+ */
+const markPluginAsActive = (player, name) => {
+  player[PLUGIN_CACHE_KEY] = player[PLUGIN_CACHE_KEY] || {};
+  player[PLUGIN_CACHE_KEY][name] = true;
+};
+
+/**
+ * Takes a basic plugin function and returns a wrapper function which marks
+ * on the player that the plugin has been activated.
+ *
+ * @private
+ * @param  {string} name
+ *         The name of the plugin.
+ *
+ * @param  {Function} plugin
+ *         The basic plugin.
+ *
+ * @return {Function}
+ *         A wrapper function for the given plugin.
+ */
+const createBasicPlugin = (name, plugin) => function() {
+  const instance = plugin.apply(this, arguments);
+
+  markPluginAsActive(this, name);
+
+  // We trigger the "pluginsetup" event on the player regardless, but we want
+  // the hash to be consistent with the hash provided for class-based plugins.
+  // The only potentially counter-intuitive thing here is the `instance` is the
+  // value returned by the `plugin` function.
+  this.trigger('pluginsetup', {name, plugin, instance});
+  return instance;
+};
+
+/**
+ * Takes a plugin sub-class and returns a factory function for generating
+ * instances of it.
+ *
+ * This factory function will replace itself with an instance of the requested
+ * sub-class of Plugin.
+ *
+ * @private
+ * @param  {string} name
+ *         The name of the plugin.
+ *
+ * @param  {Plugin} PluginSubClass
+ *         The class-based plugin.
+ *
+ * @return {Function}
+ *         A factory function for the plugin sub-class.
+ */
+const createPluginFactory = (name, PluginSubClass) => {
+
+  // Add a `name` property to the plugin prototype so that each plugin can
+  // refer to itself by name.
+  PluginSubClass.prototype.name = name;
+
+  return function(...args) {
+    this[name] = new PluginSubClass(...[this, ...args]);
+    return this[name];
+  };
+};
+
+class Plugin {
+
+  /**
+   * Plugin constructor.
+   *
+   * Subclasses should call `super` to ensure plugins are properly initialized.
+   *
+   * @param {Player} player
+   *        A Video.js player instance.
+   */
+  constructor(player) {
+    this.player = player;
+
+    if (this.constructor === Plugin) {
+      throw new Error('Plugin must be sub-classed; not directly instantiated');
+    }
+
+    // Make this object evented, but remove the added `trigger` method so we
+    // use the prototype version instead.
+    evented(this);
+    delete this.trigger;
+
+    stateful(this, this.constructor.defaultState);
+    markPluginAsActive(player, this.name);
+
+    // Bind all plugin prototype methods to this object.
+    Object.keys(Plugin.prototype).forEach(k => {
+      if (typeof this[k] === 'function') {
+        this[k] = Fn.bind(this, this[k]);
+      }
+    });
+
+    this.on('statechanged', this.handleStateChanged);
+    player.one('dispose', this.dispose);
```

Comment:
Yeah. I mean, if a player is disposed, it's "gone" (for all intents and purposes). I don't think it makes a difference if it's `one` or `on`. Maybe I'll just use `on` for simplicity, since the listener will get removed anyway when the player is disposed.